pkg/proxy/nftables/README.md

```diff
@@ -0,0 +1,103 @@
+a single packet can go `input` -> "local process" -> `output`, which it cannot. Note also
+that the `ingress` and `egress` hooks are special and mostly not available to us;
+kube-proxy lives in the middle section of diagram, with the five main netfilter hooks.
+
+There are three paths through the diagram, called the "input", "forward", and "output"
+paths, depending on which of those hooks it passes through. Packets coming from host
+network namespace processes always take the output path, while packets coming in from
+outside the host network namespace (whether that's from an external host or from a pod
+network namespace) arrive via `ingress` and take the input or forward path, depending on
+the routing decision made after `prerouting`; packets destined for an IP which is assigned
+to a network interface in the host network namespace get routed along the input path;
+anything else (including, in particular, packets destined for a pod IP) gets routed along
+the forward path.
+
+== kube-proxy's use of nftables hooks
+
+Kube-proxy uses nftables for four things:
+
+  - Using DNAT to rewrite traffic from service IPs (cluster IPs, external IPs, load balancer
+    IP, and NodePorts on node IPs) to the corresponding endpoint IPs.
+
+  - Using SNAT to masquerade traffic as needed to ensure that replies to it will come back
+    to this node/namespace (so that they can be un-DNAT-ed).
+
+  - Dropping packets that are filtered out by the `LoadBalancerSourceRanges` feature.
+
+  - Dropping packets for services with `Local` traffic policy but no local endpoints.
+
+  - Rejecting packets for services with no local or remote endpoints.
+
+This is implemented as follows:
+
+  - We do the DNAT for inbound traffic in `prerouting`: this covers traffic coming from
+    off-node to all types of service IPs, and traffic coming from pods to all types of
+    service IPs. (We *must* do this in `prerouting`, because the choice of endpoint IP may
+    affect whether the packet then gets routed along the input path or the forward path.)
+
+  - We do the DNAT for outbound traffic in `output`: this covers traffic coming from
+    host-network processes to all types of service IPs. Regardless of the final
+    destination, the traffic will take the "output path". (In the case where a
+    host-network process connects to a service IP that DNATs it to a host-network endpoint
+    IP, the traffic will still initially take the "output path", but then reappear on the
+    "input path".)
+
+  - `LoadBalancerSourceRanges` firewalling has to happen before service DNAT, so we do
+    that on `prerouting` and `output` as well, with a lower (i.e. more urgent) priority
+    than the DNAT chains.
+
+  - The `drop` and `reject` rules for services with no endpoints don't need to happen
+    explicitly before or after any other rules (since they match packets that wouldn't be
+    matched by any other rules). But with kernels before 5.9, `reject` is not allowed in
```

Comment:
Since the comment below is restored, should we change it to "... because reject action is only valid in chains using the input, forward or output hooks before kernel 5.9" to avoid being misleading?
https://github.com/kubernetes/kubernetes/blob/773ad73b90c2836685358a0fafe285b9e4cdcc19/pkg/proxy/nftables/proxier.go#L349-L351

Reply:
I don't think it's really misleading... both statements are true

Reply:
I didn't mean the doc in this PR but the code comment above merged via https://github.com/kubernetes/kubernetes/pull/122614 could be misleading without "before kernel 5.9".
But it's not the issue of this PR anyway.

Reply:
ohhh, I see. Yeah, we should fix that at some point...